src/shell/scripts/omp.zsh

```diff
@@ -143,20 +154,43 @@ function enable_poshtooltips() {
   bindkey " " _posh-tooltip
 }
 
+# Helper function for posh::decorate_widget
+# It calls the posh function right after the original definition of the widget
+# $1 is the name of the widget to call
+# $2 is the posh widget name
+posh::call_widget()
+{
+  builtin zle "${1}" &&
+  ${2}
+}
+
+# decorate_widget
+# Allows to preserve any user defined value that may have been defined before posh tries to redefine it.
+# Instead, we keep the previous function and decorate it with the posh additions
+# $1: The name of the widget to decorate
+# $2: The name of the posh function to decorate it with
+function posh::decorate_widget() {
+  typeset -F SECONDS
+  local prefix=orig-s$SECONDS-r$RANDOM # unique each time, in case we're sourced more than once
+  cur_widget=${1}
+  posh_widget=${2}
+
+  case ${widgets[$cur_widget]:-""} in
+    # Already decorated: do nothing.
+    user:_posh-decorated-*);;
+
+    user:*)
+      zle -N $prefix-$cur_widget ${widgets[$cur_widget]#*:}
+      eval "_posh-decorated-${(q)prefix}-${(q)cur_widget}() { posh::call_widget ${(q)prefix}-${(q)cur_widget} ${(q)posh_widget} -- \"\$@\" }"
+      zle -N $cur_widget _posh-decorated-$prefix-$cur_widget;;
+
+    # For now, do not decorate if it's not a user:*
```

Comment:
@lewis-yeung so this should in fact just add it, right?

Reply:
I need to check with ${+widgets[$cur_widget]} if the original widget exists or not.
I will carve out some time tonight for this + testing the bindkey removal.

Reply:
> @lewis-yeung so this should in fact just add it, right?

@JanDeDobbeleer Yes. And I wonder if we can utilize the decorating logic to resolve #5079. Will try a fix when I'm available.

Reply:
> I wonder if we can utilize the decorating logic to resolve #5079.

Seems unrelated. 🤷🏻‍♂️

Reply:
Yes, that's about key bindings. More tricky that one I guess (or meeds a different approach).

Reply:
> I wonder if we can utilize the decorating logic to resolve #5079.

@JanDeDobbeleer Exciting news: this is practical with the help of decorating a widget. I've just tested out a fix and it resolved #5079. I'll double check and make a PR in a few days.